agent/app/agent_unix.go

```diff
@@ -0,0 +1,76 @@
+// +build linux
+
+// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package app
+
+import (
+	"os"
+	"os/signal"
+	"syscall"
+
+	log "github.com/cihub/seelog"
+	"golang.org/x/net/context"
+)
+
+// startSigchldHandler registers a channel to receiev SIGCHLD signals for
+// the agent process. On receiving SIGCHLD, it invokes the wait4 syscall
+// so that child processes are appropriately cleaned up.
+func (agent *ecsAgent) startSigchldHandler(ctx context.Context) {
+	signals := make(chan os.Signal)
+	signal.Notify(signals, syscall.SIGCHLD)
+	go processSignal(ctx, signals)
+}
+
+func processSignal(ctx context.Context, signals <-chan os.Signal) {
+	log.Info("Starting the SIGCHLD handler")
+	for {
+		select {
+		case s := <-signals:
+			log.Debugf("Received SIGCHLD: %s", s.String())
+			go wait()
+		case <-ctx.Done():
+			log.Info("Stopping the SIGCHLD handler")
+			return
+		}
+	}
+}
+
+// wait wraps the Wait4 syscall and returns the error if any
+func wait() {
+	var ws syscall.WaitStatus
+	var ru syscall.Rusage
+	// More information on wait4 syscall can be found in manual pages
+	// via `man 2 wait4` command. The syscall is used to wait for state
+	// changes in the child processes of the Agent. Examples of such
+	// processes include CNI plugins and the dhclient processes started
+	// by these plugins. As per man pages, if a wait is not performed,
+	// then the terminated child remains in a "zombie" state, which is
+	// especially problematic for the dhclient process as the network
+	// resources, including the namespace that holds the ENI would not
+	// be properly cleaned up.
+	//
+	// wait4(pid, status, options, rusage) is equivalent to
+	// the waitpid(pid, status, options) syscall. The value of -1 for
```

Comment:
Why use `wait4(2)` instead of `waitpid(2)`?  When I look at the manual for `wait4`, it says "These functions are obsolete; use waitpid(2) or waitid(2) in new programs."